lib/charms/mongodb/v0/set_status.py

```diff
@@ -150,4 +161,99 @@ def is_unit_status_ready_for_upgrade(self) -> bool:
 
         return False
 
+    def process_statuses(self) -> StatusBase:
+        """Retrieves statuses from processes inside charm and returns the highest priority status.
+
+        When a non-fatal error occurs while processing statuses, the error is processed and
+        returned as a statuses.
+
+        TODO: add more status handling here for other cases: i.e. TLS, or resetting a status that
+        should not be reset
+        """
+        # retrieve statuses of different services running on Charmed MongoDB
+        deployment_mode = (
+            "replica set" if self.charm.is_role(Config.Role.REPLICATION) else "cluster"
+        )
+        waiting_status = None
+        try:
+            statuses = self.get_statuses()
+        except OperationFailure as e:
+            if e.code in [UNAUTHORISED_CODE, AUTH_FAILED_CODE]:
+                waiting_status = f"Waiting to sync passwords across the {deployment_mode}"
+            elif e.code == TLS_CANNOT_FIND_PRIMARY:
+                waiting_status = (
+                    f"Waiting to sync internal membership across the {deployment_mode}"
+                )
+            else:
+                raise
```

Comment:
Another suggestion which will make this code easier to update in the future if we have to handle more status codes:
```suggestion
            match e.code:
                case UNAUTHORISED_CODE | AUTH_FAILED_CODE:
                    waiting_status = f"Waiting to sync passwords across the {deployment_mode}"
                case TLS_CANNOT_FIND_PRIMARY:
                    waiting_status = (
                        f"Waiting to sync internal membership across the {deployment_mode}"
                    )
                case _:
                    raise
```